docs/arcadia-users-group/20231128-testing-datasets/lesson.md

```diff
@@ -0,0 +1,289 @@
+# Building test data sets
+
+In our previous lesson, we covered different types of tests.
+In the unit test section, we used five functions to test that our `mean()` function produced the output that we expected it to.
+For each of these functions, we started by defining our test data, `num_list`, a list of numbers for which we wanted to calculate the mean.
+
+These test data were small; each list only had five numbers in it.
+The "correct" answer to the test was clear with relatively little human effort (we could take a few seconds to calculate the mean of each list in our own heads).
+The creation of each test data set was encoded in the test itself.
+
+Below we discuss strategies for creating test data sets to help maintain some of these desirable qualities in real code bases.
+
+## Lesson set up
+
+### Software installation
+
+This lesson will take advantage of the skills we've learned in many previous lessons.
+We'll use [Jupyter Notebooks](https://training.arcadiascience.com/arcadia-users-group/20221024-jupyter-notebooks/lesson/), [GitHub](https://training.arcadiascience.com/workshops/20220920-intro-to-git-and-github/lesson/), [conda](https://training.arcadiascience.com/arcadia-users-group/20221017-conda/lesson/), and [Python](https://training.arcadiascience.com/arcadia-users-group/20230228-intro-to-python-1/lesson/).
+This is more overhead than we typically strive for in a lesson, but we hope that it's a chance to practice these skills to achieve a new goal.
+In the future, we may provide a [GitPod](https://www.gitpod.io/) environment for learners to use while working through this lesson, however if possible we would prefer to empower users to start implementing tests on their own computers using their own setup.
+
+To start this lesson, we'll begin by creating a conda environment that has the tools we'll need.
+
+```
+mamba create -n augtest2 jupyter pytest pandas matplotlib seaborn pillow
+conda activate augtest2
+```
+
+We'll also create a folder to help us stay organized.
+```
+mkdir 20231128-aug-testing-lesson
+cd 20231128-aug-testing-lesson
+```
+
+Once our environment is activated, start a Jupyter notebook
+
+```
+jupyter notebook
+```
+
+### Example function
+
+As a running example, similar to the `mean()` function we used in the previous lesson, we'll use `distribution()`, a simple function that returns the distribution of characters in text.
+This function is intentionally implemented inefficiently, to emulate an expensive function of the type typically found in computational biology workflows, that would require testing.
```

Comment:
@giladmishne can you comment on this? what is inefficient about it, or can we remove that detail? I think as phrased it makes it sound like only inefficient things need testing

Reply:
Ah, I can see why this is confusing. How about (don't want to get into complexity calculations too much):

```
As an aside, this function is intentionally implemented inefficiently; the same functionality can be implemented much faster in Python. We do this to later demonstrate what to do when a complex function -- of the type typically found in computational biology workflows -- is too slow to run on longer test inputs.

If you're curious, the reason `distribution()` is inefficient is that an input text of length _n_ will be read _n^2_ times -- for each character in the text, `count()` is called, and it will in turn traverse the entire input. Instead, a single pass over the text can be used to keep running counts of the characters observed.
```

Or we can just remove the sentence on line 44 in the original text.

Reply:
no i love this explanation! this is great, thank you. i'll add it in.